### ISSUE TITLE
Add mixed precision training support

### ISSUE THREAD

**Issue by MaxGhenis** [PolicyEngine/L0#21]
## Description
Add support for Automatic Mixed Precision (AMP) training with L0 regularization.

## Motivation
Mixed precision can significantly speed up training and reduce memory usage, especially important when adding L0 gates increases model size.

## Implementation Requirements
```python
from torch.cuda.amp import autocast, GradScaler

# Ensure L0 gates work with AMP
class L0Linear(nn.Module):
    @autocast()
    def forward(self, x):
        # Ensure gate computations handle fp16/fp32 correctly
        gates = self.weight_gates().float()  # Gates in fp32
        masked_weight = self.weight * gates
        return F.linear(x, masked_weight, self.bias)

# Training example
scaler = GradScaler()
for x, y in loader:
    with autocast():
        output = model(x)
        loss = criterion(output, y) + lambda_l0 * model.get_l0_loss()
    
    scaler.scale(loss).backward()
    scaler.step(optimizer)
    scaler.update()
```

## Testing
- Verify numerical stability with fp16
- Compare convergence with/without AMP
- Check gradient scaling doesn't affect sparsity learning
- Benchmark speedups

## Considerations  
- Gates might need to stay in fp32 for stability
- Log-odds computations can be sensitive to precision
- Temperature scheduling interaction with gradient scaling
